[mpm-itk] Switching vhosts during the same connection in Debian
Steinar H. Gunderson sgunderson at bigfoot.com
Sat Jul 31 01:55:44 CEST 2010

On Fri, Jul 23, 2010 at 06:26:11PM +0200, Steinar H. Gunderson wrote:
> I reiterate that the best mpm-itk fix would probably be to make “cannot read
> .htaccess after uid switch” behave identically as “cannot setuid after uid
> switch”, ie. close the HTTP connection straight. I have no idea how easy it
> is to do, though.

You can try this patch as a first approximation. The logging doesn't actually
appear to work, it's pretty much untested, and probably needs some more
sanity checking, but it just might help with your issue:

--- a/server/config.c	2010-07-21 20:11:07.000000000 +0200
+++ b/server/config.c	2010-07-31 01:53:15.000000000 +0200
@@ -1840,6 +1867,15 @@
         else {
             if (!APR_STATUS_IS_ENOENT(status)
                 && !APR_STATUS_IS_ENOTDIR(status)) {
+#ifdef ITK_MPM
+               if (getuid() != 0) {
+                       ap_log_error(APLOG_MARK, APLOG_WARNING, status, r, 
+                           "Couldn't read %s, closing connection.",
+                           filename);
+                       ap_lingering_close(r->connection);
+                       exit(0);
+               }
+#endif
                 ap_log_rerror(APLOG_MARK, APLOG_CRIT, status, r,
                               "%s pcfg_openfile: unable to check htaccess file, "
                               "ensure it is readable",

/* Steinar */
-- 
Homepage: http://www.sesse.net/

